pkg/resources/manager/manager.go

```diff
@@ -0,0 +1,561 @@
+	var count int64 = math.MaxInt64
+	for key, classQuantity := range class.resources {
+		available, err := getResource(key, r.resourcesAvailable)
+		if err != nil {
+			return err
+		}
+
+		newCount := int64(math.Floor(float64(available.Value()) / float64(classQuantity.Value())))
+		if newCount < count {
+			count = newCount
+		}
+	}
+
+	if count < 0 {
+		count = 0
+	}
+
+	class.available = count
+
+	return nil
+}
+
+func (r *resourceManager) modifyResources(resources core.ResourceList) error {
+	for name, source := range r.sources {
+		err := source.Modify(resources)
+		if err != nil {
+			return fmt.Errorf("source %s couldn't modified resources: %w", name, err)
+		}
+	}
+
+	return nil
+}
+
+// checkContext will return immediately error without additional allocate/deallocate operation
+// when parent context is closed. It optimizes time spend in mutex
+func (r *resourceManager) checkContext() error {
+	var err error
+	if r.ctx.Err() != nil {
+		err = fmt.Errorf("context error: %w", r.ctx.Err())
+		r.operationError = err
+	}
+
+	return err
+}
+
+func getResource(name core.ResourceName, resources core.ResourceList) (*resource.Quantity, error) {
+	quantity, ok := resources[name]
+	if !ok {
+		return nil, fmt.Errorf("failed to get %s resource: %w", name, ErrResourceUnsupported)
+	}
+
+	return &quantity, nil
+}
+
+func (r *resourceManager) getMachineClass(name string) (*machineclass, error) {
+	for _, class := range r.machineClasses {
+		if class.name == name {
+			return class, nil
+		}
+	}
+
+	return nil, ErrMachineClassMissing
+}
+
+// printAvailableResources is helpful function for avoid unnecessary operations
+func (r *resourceManager) printAvailableResources(operation string) {
+	const traceLevel = 2
+	if !r.log.V(traceLevel).Enabled() {
+		return
+	}
+
+	r.log.V(traceLevel).Info("Available resources after " + operation + ": " + r.convertResourcesToString(r.resourcesAvailable))
+	r.log.V(traceLevel).Info("Machineclasses availibility: " + r.getMachineClassAvailibilityAsString())
+}
+
+func (r *resourceManager) convertResourcesToString(resources core.ResourceList) string {
+	var result string
+	type res struct {
+		name     string
+		quantity resource.Quantity
+	}
+
+	arr := make([]res, 0, len(resources))
+
+	for key, quantity := range resources {
+		arr = append(arr, res{name: string(key), quantity: quantity})
+	}
+
+	sort.Slice(arr, func(i, j int) bool {
+		return arr[i].name < arr[j].name
+	})
+
+	for index := range arr {
+		result += arr[index].name + ": " + arr[index].quantity.String() + sep
+	}
+
+	return removeSeparatorFromEnd(result)
+}
+
+func (r *resourceManager) getMachineClassAvailibilityAsString() string {
+	var result string
+	for _, class := range r.machineClasses {
+		result += class.name + ": " + strconv.FormatInt(class.available, 10) + sep
+	}
+
+	return removeSeparatorFromEnd(result)
+}
+
+// reset internal state of manager and allow reinit
+// Use it for unit test only.
+func (r *resourceManager) reset() {
+	mng.ctx = nil
```

Comment:
i checked logic again and this solution is better. Main problem with your suggestion is reference values as sources, or available resources. Their default value is nil. So if i create only new object, it contains nil and i will have to extend logic in another functions for this specific case. Reset function solving this problem for me because it initialize reference types.